File: Member 2/packages/contracts/src/trace.ts
import { z } from "zod";
import { EpochMs, ErrorRecord, Sha256, ShortText, externalObject } from "./common";
import { AdapterId, ArtifactId, EventId, RunId, SpanId, TaskId } from "./ids";

export const ExecutionLocation = z.enum(["local_cpu", "local_gpu", "workers_ai", "remote_provider", "tool_sandbox"]);
export const SpanStatus = z.enum(["queued", "running", "succeeded", "failed", "blocked", "cancelled"]);
export const ModelIdentity = z.object({
  baseModel: ShortText,
  baseRevision: ShortText,
  baseSha256: Sha256,
  quantization: ShortText,
  adapterId: AdapterId.optional(),
  adapterVersion: ShortText.optional(),
  adapterSha256: Sha256.optional(),
  adapterScale: z.number().positive().max(100).optional(),
  trainingRunId: z.string().max(128).optional(),
  datasetManifestSha256: Sha256.optional(),
  promptVersion: ShortText,
  agentVersion: ShortText,
}).strict();

export const ToolCall = z.object({
  tool: ShortText,
  purpose: z.string().min(1).max(1_024),
  status: z.enum(["allowed", "denied", "succeeded", "failed"]),
  latencyMs: z.number().int().nonnegative().max(86_400_000),
}).strict();

export const TraceEvent = externalObject({
  eventId: EventId,
  sequence: z.number().int().positive(),
  taskId: TaskId,
  runId: RunId,
  spanId: SpanId,
  parentSpanId: SpanId.optional(),
  nodeId: z.string().regex(/^node_[A-Za-z0-9_-]{1,80}$/),
  agent: ShortText,
  model: ModelIdentity,
  promptHash: Sha256,
  inputArtifactRefs: z.array(ArtifactId).max(200),
  outputArtifactRef: ArtifactId.optional(),
  tokensIn: z.number().int().nonnegative(),
  tokensOut: z.number().int().nonnegative(),
  costUsd: z.number().nonnegative(),
  costCloudEquivalentUsd: z.number().nonnegative(),
  latencyMs: z.number().int().nonnegative(),
  executionLocation: ExecutionLocation,
  toolCalls: z.array(ToolCall).max(100),
  status: SpanStatus,
  verdict: z.enum(["pass", "revise", "blocked"]).optional(),
  error: ErrorRecord.optional(),
  startedAt: EpochMs,
  finishedAt: EpochMs.optional(),
});

export type TraceEvent = z.infer<typeof TraceEvent>;
export type ModelIdentity = z.infer<typeof ModelIdentity>;

File: krishang/src/helios/contracts/trace.py
from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import Field

from helios.clock import now_utc
from helios.ids import new_id
from .common import WireModel


class RedactionLevel(StrEnum):
    PUBLIC = "public"
    INTERNAL = "internal"
    PRIVATE = "private"
    SECRET_REDACTED = "secret_redacted"


class CanonicalEvent(WireModel):
    schema_version: str = "1.0"
    event_id: str = Field(default_factory=lambda: new_id("event"))
    type: str
    source: str = "helios-runtime"
    timestamp: datetime = Field(default_factory=now_utc)
    sequence: int = 0
    task_id: str | None = None
    run_id: str | None = None
    span_id: str | None = None
    label: Literal["live", "dry-run", "degraded", "replayed", "fixture"] = "live"
    payload: dict[str, Any] = Field(default_factory=dict)
    redaction_level: RedactionLevel = RedactionLevel.INTERNAL


class Span(WireModel):
    schema_version: str = "1.0"
    span_id: str = Field(default_factory=lambda: new_id("span"))
    run_id: str
    task_id: str
    parent_span_id: str | None = None
    node_id: str
    agent: str
    agent_version: str = "1.0"
    model: str = "deterministic"
    prompt_hash: str = ""
    input_artifact_refs: list[str] = Field(default_factory=list)
    output_artifact_ref: str | None = None
    tokens_in: int = 0
    tokens_out: int = 0
    cost_usd: float = 0
    cost_cloud_equiv_usd: float = 0
    latency_ms: float = 0
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    verdict: str | None = None
    error: str | None = None
    degraded: bool = False
